### PR TITLE
[FLINK-19186][python] Add Python building blocks to make sure the basic functionality of Pandas Batch Group Aggregation could work

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will add Python building blocks to make sure the basic functionality of Pandas Batch Group Aggregation could work*

## Brief change log

  - *Add Aggregation interface*
  - *Add udaf decorator*
  - *Add PandasAggregateFunctionOperation*

## Verifying this change

This change added tests and can be verified as follows:

  - *Add IT test test_group_aggregate_function in test_pandas_udaf.py*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/udf.py [hunk 1]
```diff
@@ -147,6 +234,41 @@ def eval(self, *args):
...
+        return accumulator[0]
+
+    def create_accumulator(self):
+        return []
+
+    def accumulate(self, accumulator, *args):
+        accumulator.append(self.func(*args))
+
+
+class WrapperPandasAggregateFunction(object):
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:256]: Rename it as `PandasAggregateFunctionWrapper` and add some description about this class?

File: flink-python/pyflink/table/udf.py [hunk 2]
```diff
@@ -313,6 +435,70 @@ def _create_judtf(self):
...
+                return JPythonFunctionKind.PANDAS
+            else:
+                raise TypeError("Unsupported udf_type: %s." % udf_type)
+
+        func = self._func
+        if not isinstance(self._func, UserDefinedFunction):
+            func = DelegatingPandasAggregateFunction(self._func)
+
+        if self._udaf_type == "pandas":
+            func = WrapperPandasAggregateFunction(func)
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:476]: We can also create the wrapper class during execution.

File: flink-python/pyflink/table/udf.py [hunk 3]
```diff
@@ -313,6 +435,70 @@ def _create_judtf(self):
         return j_table_function
 
 
+class UserDefinedAggregateFunctionWrapper(UserDefinedFunctionWrapper):
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:438]: Currently we have UserDefinedAggregateFunctionWrapper, UserDefinedTableFunctionWrapper, UserDefinedScalarFunctionWrapper, it seems to me that most of the implementation is duplicate. Could we refactor them a bit?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonAggregateFunction.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	@Override
+	public TypeInformation getResultType() {
+		return resultType;
+	}
+
+	@Override
+	public TypeInformation getAccumulatorType() {
+		return resultType;
```
Comment (dianfu) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonAggregateFunction.java:109]: Why returns the resultType?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecPythonGroupAggregate.scala [hunk 5]
```diff
@@ -168,6 +177,78 @@ class BatchExecPythonGroupAggregate(
...
+        ret, getPythonWorkerMemory(planner.getTableConfig.getConfiguration).getBytes)
+    }
+    ret
+  }
+
+  private[this] def createPythonOneInputTransformation(
+      inputTransform: Transformation[RowData],
+      inputRowType: RowType,
+      outputRowType: RowType,
+      groupingSet: Array[Int],
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecPythonGroupAggregate.scala:202]: groupingSet is never used?

**Review by dianfu — CHANGES_REQUESTED** [review]
@HuangXingBo Thanks a lot for the PR. There is only one test case for such a big PR. Could we add more test cases, e.g.
- Add tests for TableEnvironment.register_function and create_temporary_function for AggregateFunction
- Add tests which have auxGrouping
- Add tests which have multiple pandas udaf
- Add tests for pandas udaf which takes multiple arguments as input and one of them is constant?
- Add tests for pandas UDAF which extends AggregateFunction?
- Add tests of Pandas UDAF mixing with Pandas UDF, non-pandas UDF?
- Add tests without groupby clause?

Besides, the documentation is not added in this PR. Could we create a JIRA to track that if we want to add the documentation in another PR?

File: flink-python/pyflink/fn_execution/beam/beam_operations_fast.pyx [hunk 6]
```diff
@@ -173,6 +173,28 @@ cdef class DataStreamStatelessFunctionOperation(BeamStatelessFunctionOperation):
...
+    def generate_func(self, udfs):
+        pandas_functions, variable_dict, user_defined_funcs = reduce(
+            lambda x, y: (
+                ','.join([x[0], y[0]]),
+                dict(chain(x[1].items(), y[1].items())),
+                x[2] + y[2]),
+            [operation_utils.extract_user_defined_function(udf) for udf in udfs])
+        variable_dict['wrap_pandas_result'] = operation_utils.wrap_pandas_result
+        mapper = eval('lambda value: wrap_pandas_result([%s])' % pandas_functions, variable_dict)
+        if self._is_python_coder:
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/beam/beam_operations_fast.pyx:191]: We currently only provide the implementation of arrow coder with python version, so it is always true

File: flink-python/pyflink/table/tests/test_pandas_udaf.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        self.t_env.register_table_sink("Results", table_sink)
+        t.group_by("a") \
+            .select(t.a, mean_udaf(t.b)) \
+            .execute_insert("Results") \
+            .wait()
+        actual = source_sink_utils.results()
+        self.assert_equals(actual, ["1,5.0", "2,2.0", "3,2.0"])
+
+
+@udaf(result_type=DataTypes.FLOAT(), udaf_type="pandas")
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/table/tests/test_pandas_udaf.py:46]: Yes.Make sense.

File: flink-python/pyflink/table/udf.py (hunk 3 above)
Comment (HuangXingBo) [reply flink-python/pyflink/table/udf.py:438]: Yes. Make sense

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

File: flink-python/pyflink/table/udf.py [hunk 8]
```diff
@@ -312,6 +422,54 @@ def _create_judtf(self):
...
+            bytearray(serialized_func),
+            j_input_types,
+            j_result_type,
+            j_accumulator_type,
+            j_function_kind,
+            self._deterministic,
+            _get_python_env())
+        return j_aggregate_function
+
+    def _create_delegate_function(self) -> UserDefinedFunction:
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:470]: Add an assert that the function_type should be pandas?

File: flink-python/pyflink/table/udf.py [hunk 9]
```diff
@@ -312,6 +422,54 @@ def _create_judtf(self):
...
+                "Invalid returnType: returnType should be DataType but is {}".format(result_type))
+        if accumulator_type is not None and not isinstance(accumulator_type, DataType):
+            raise TypeError(
+                "Invalid accumulator_type: accumulator_type should be DataType but is {}".format(
+                    accumulator_type))
+        self._result_type = result_type
+        self._accumulator_type = accumulator_type
+
+    def _create_judf(self, serialized_func, j_input_types, j_function_kind):
+        if self._function_type == "pandas" and not self._accumulator_type:
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:449]:
```suggestion
        if self._function_type == "pandas":
```

File: flink-python/pyflink/table/udf.py [hunk 10]
```diff
@@ -421,3 +586,51 @@ def udtf(f=None, input_types=None, result_types=None, deterministic=None, name=N
...
+            ...     return v.mean()
+
+    :param f: user-defined aggregate function.
+    :type f: function or UserDefinedFunction or type
+    :param input_types: optional, the input data types.
+    :type input_types: list[DataType] or DataType
+    :param result_type: the result data type.
+    :type result_type: DataType
+    :param accumulator_type: the accumulator data type.
+    :type accumulator_type: DataType
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:611]: Use typehint if possible?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the update. Only a few minor comments.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

File: flink-python/pyflink/table/udf.py [hunk 11]
```diff
@@ -421,3 +600,53 @@ def udtf(f=None, input_types=None, result_types=None, deterministic=None, name=N
...
+    Example:
+        ::
+
+            >>> # The input_types is optional.
+            >>> @udaf(result_type=DataTypes.FLOAT(), func_type="pandas")
+            ... def mean_udaf(v):
+            ...     return v.mean()
+
+    :param f: user-defined aggregate function.
+    :type f: function or UserDefinedFunction or type
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:621]: can be removed

File: flink-python/pyflink/table/udf.py [hunk 12]
```diff
@@ -361,26 +528,38 @@ def udf(f=None, input_types=None, result_type=None, deterministic=None, name=Non
...
+                    (default: general)
     :type udf_type: str
     :return: UserDefinedScalarFunctionWrapper or function.
     :rtype: UserDefinedScalarFunctionWrapper or function
 
     .. versionadded:: 1.10.0
     """
-    if udf_type not in ('general', 'pandas'):
-        raise ValueError("The udf_type must be one of 'general, pandas', got %s." % udf_type)
+    if func_type not in ('general', 'pandas'):
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:542]: the warning should be moved after
```
    if udf_type:
        import warnings
        warnings.warn("The param udf_type is deprecated in 1.12. Use func_type instead.")
        func_type = udf_type
```

File: flink-python/pyflink/table/udf.py [hunk 13]
```diff
@@ -421,3 +600,53 @@ def udtf(f=None, input_types=None, result_types=None, deterministic=None, name=N
                                  deterministic=deterministic, name=name)
     else:
         return _create_udtf(f, input_types, result_types, deterministic, name)
+
+
+def udaf(f: Union[Callable, UserDefinedFunction, Type] = None,
+         input_types: Union[List[DataType], DataType] = None, result_type: DataType = None,
+         accumulator_type: DataType = None, deterministic: bool = None, name: str = None,
+         func_type: str = "pandas") -> Union[UserDefinedAggregateFunctionWrapper, Callable]:
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:608]: What about making the default value `general`?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks for the update. There are still a few minor comments. If you agree with them, I'll integrate the comments during merging the PR.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I agree with the comments.

**Review by dianfu — APPROVED** [review]
@HuangXingBo Thanks for the great work! LGTM. Will merge once the tests passed.

**Pull request closed without merging**